clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -472,6 +472,12 @@ private static int parseAcks(String acksString) {
             // handling exceptions and record the errors;
             // for API exceptions return them in the future,
             // for other exceptions throw directly
+        } catch (TimeoutException e) {
+            this.errors.record();
+            this.metrics.sensor("buffer-exhausted-records").record();
```

Comment:
Are all cases of this exception due to buffer exhaustion? I thought that it could also happen in other cases? One option would be to keep `BufferExhaustedException` and have it inherit from `TimeoutException`. Thoughts?